_posts/2017-11-24-upgrade-rails-from-4-0-to-4-1.markdown

```diff
@@ -0,0 +1,226 @@
+your [Ruby on Rails](http://rubyonrails.org/) application from [version 4.0](http://guides.rubyonrails.org/4_0_release_notes.html) to [4.1](http://guides.rubyonrails.org/4_1_release_notes.html).
+
+<!--more-->
+
+1. [Ruby version](#ruby-version)
+2. [Gems](#gems)
+3. [Config files (config/)](#config-files)
+4. [Application code](#application-code)
+  1. [Models (app/models/)](#models)
+5. [Tests](#tests)
+6. [Miscellaneous](#miscellaneous)
+7. [Next steps](#next-steps)
+
+<h2 id="ruby-version">1. Ruby version</h2>
+
+Rails 4.1 requires Ruby 1.9.3 or later. Ruby 1.8.7 support was dropped in
+Rails 4.0, so you should already be running 1.9.3 or later. For Rails 4.1,
+Ruby 2.0 (or newer) is preferred according to the [official upgrade guide](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#ruby-versions).
+
+<h2 id="gems">2. Gems</h2>
+
+If your application relies on `MultiJSON`, you will need to add the gem to your
+`Gemfile` (`gem 'multi_json'`) if it's not already there, since it was removed
+from Rails 4.1.
+
+Alternatively, stop using MultiJSON and migrate your application to use
+`to_json` and `JSON.parse`.
+
+<h2 id="config-files">3. Config files</h2>
+
+Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task).
+You can use this task as a guideline as explained thoroughly in
+[this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html).
+
+As an alternative, check out [RailsDiff](http://railsdiff.org/4.0.13/4.1.16),
+which provides an overview of the changes in a basic Rails app between 4.0 and
+4.1 (or any other source/target versions).
+
+<h2 id="application-code">4. Application code</h2>
+
+<h2 id="models">a. Models</h2>
+
+- Return from callbacks is no longer allowed:
+
+Before:
+```ruby
+before_save { return false }
+```
+
+After:
+```ruby
+before_save { false }
+```
+
+See: https://github.com/rails/rails/pull/13271
+
+- Removal of deprecated finders:
+
+`activerecord-deprecated_finders` (https://github.com/rails/activerecord-deprecated_finders)
+was removed as a dependency from Rails 4.1. From the gem's README:
+
+```
+To migrate dynamic finders to Rails 4.1+:
+
+find_all_by_... should become where(...).
+find_last_by_... should become where(...).last.
+scoped_by_... should become where(...).
+find_or_initialize_by_... should become find_or_initialize_by(...).
+find_or_create_by_... should become find_or_create_by(...).
+```
+
+If you can't afford to upgrade the finders now, then add the gem back yourself
+into the `Gemfile`:
+
+```ruby
+gem 'activerecord-deprecated_finders'
+```
+
+See our [last upgrade post](https://www.ombulabs.com/blog/rails/upgrades/upgrade-rails-from-3-2-to-4-0.html) for more information.
+
+- Default scopes are now chained to other scopes:
+
+If you thought default scopes on models could be confusing, there's even another
+(un?)expected twist to it:
```

Comment:
That was intentional, since `default_scope` historically has unexpected behavior. 

It's up to the reader to decide whether it's expected or unexpected: 
- Unexpected because unless you're told about it, you wouldn't expect this to be the case when going from 4.0 to 4.1. 
- Expected because you are bound to run into some trouble when using `default_scope`.